tests/benchmarks/cpu/benchmark_propagator.cpp

```diff
@@ -76,20 +76,47 @@ auto toy_cfg =
     toy_det_config{}.n_brl_layers(4u).n_edc_layers(7u).do_check(false);
 
 void fill_tracks(vecmem::vector<free_track_parameters<algebra_t>> &tracks,
-                 const std::size_t theta_steps, const std::size_t phi_steps) {
-    // Set momentum of tracks
-    const scalar mom_mag{10.f * unit<scalar>::GeV};
+                 const std::size_t n_tracks, bool do_sort = true) {
+    using scalar_t = dscalar<algebra_t>;
+    using uniform_gen_t =
+        detail::random_numbers<scalar_t,
+                               std::uniform_real_distribution<scalar_t>>;
+    using trk_generator_t =
+        random_track_generator<free_track_parameters<algebra_t>, uniform_gen_t>;
+
+    trk_generator_t::configuration trk_gen_cfg{};
+    trk_gen_cfg.seed(42u);
+    trk_gen_cfg.n_tracks(n_tracks);
+    trk_gen_cfg.phi_range(-constant<scalar_t>::pi, constant<scalar_t>::pi);
+    trk_gen_cfg.eta_range(-4.f, 4.f);
+    trk_gen_cfg.mom_range(1.f * unit<scalar_t>::GeV,
+                          100.f * unit<scalar_t>::GeV);
+    trk_gen_cfg.origin({0.f, 0.f, 0.f});
+    trk_gen_cfg.origin_stddev({0.f * unit<scalar_t>::mm,
+                               0.f * unit<scalar_t>::mm,
+                               0.f * unit<scalar_t>::mm});
 
     // Iterate through uniformly distributed momentum directions
-    for (auto traj : uniform_track_generator<free_track_parameters<algebra_t>>(
-             phi_steps, theta_steps, mom_mag)) {
+    for (auto traj : trk_generator_t{trk_gen_cfg}) {
         tracks.push_back(traj);
     }
+
+    if (do_sort) {
+        // Sort by theta angle
+        const auto traj_comp = [](const auto &lhs, const auto &rhs) {
+            return getter::theta(lhs.dir()) < getter::theta(rhs.dir());
```

Comment:
Are you sorting to get better performance? if so, I would do it with $|\pi/2 -\theta|$ given that toy geometry is symmetry around z=0

Reply:
That's actually a good suggestion!

Reply:
But the volumes and surfaces in $z > 0$ are - even though they are geometrically the same as the ones in $z < 0$ - still different entries in memory, right? So perhaps sorting on $\theta$ might still give us better memory locality?

Reply:
It might be worth measuring the performance for both cases. 
I think the memory locality gets beneficial only if all threads in a warp assess the same surface at the same time. Assessing the adjacent surfaces won't be very helpful given that our memory layout is full-SoA. 
In the ODD full chain benchmark, I observed that $\pi/2 - \theta$ shows better performance than $\theta$

Reply:
They might still be accessing similar grid/material bins or not? We might try spatial sorting for the non-deduplicated data at some point.

Just running the benchmark here, I can't spot a difference:
```
Running ./bin/detray_benchmark_cuda_array
Run on (48 X 1795.91 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x24)
  L1 Instruction 32 KiB (x24)
  L2 Unified 512 KiB (x24)
  L3 Unified 32768 KiB (x4)
Load Average: 0.00, 0.68, 1.91
--------------------------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------
CUDA unsync propagation/8      1612786 ns      1608145 ns          379 TracksPropagated=39.7974k/s
CUDA unsync propagation/16     1977033 ns      1971383 ns          362 TracksPropagated=129.858k/s
CUDA unsync propagation/32     2464757 ns      2457488 ns          285 TracksPropagated=416.686k/s
CUDA unsync propagation/64     2822198 ns      2813464 ns          249 TracksPropagated=1.45586M/s
CUDA unsync propagation/128    5024377 ns      5012248 ns          149 TracksPropagated=3.26879M/s
CUDA unsync propagation/256   15127111 ns     15094245 ns           46 TracksPropagated=4.34179M/s
CUDA sync propagation/8        1591341 ns      1586690 ns          400 TracksPropagated=40.3355k/s
CUDA sync propagation/16       1915189 ns      1909652 ns          367 TracksPropagated=134.056k/s
CUDA sync propagation/32       2425989 ns      2418718 ns          289 TracksPropagated=423.365k/s
CUDA sync propagation/64       2807353 ns      2798694 ns          250 TracksPropagated=1.46354M/s
CUDA sync propagation/128      4991046 ns      4979190 ns          151 TracksPropagated=3.2905M/s
CUDA sync propagation/256     15499985 ns     15467954 ns           45 TracksPropagated=4.23689M/s

Running ./bin/detray_benchmark_cuda_array
Run on (48 X 1798.5 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x24)
  L1 Instruction 32 KiB (x24)
  L2 Unified 512 KiB (x24)
  L3 Unified 32768 KiB (x4)
Load Average: 0.52, 0.67, 1.72
--------------------------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------
CUDA unsync propagation/8      1606549 ns      1601741 ns          385 TracksPropagated=39.9565k/s
CUDA unsync propagation/16     1976884 ns      1971046 ns          359 TracksPropagated=129.88k/s
CUDA unsync propagation/32     2474099 ns      2466805 ns          284 TracksPropagated=415.112k/s
CUDA unsync propagation/64     2833345 ns      2824586 ns          248 TracksPropagated=1.45012M/s
CUDA unsync propagation/128    5013851 ns      5001393 ns          149 TracksPropagated=3.27589M/s
CUDA unsync propagation/256   15183524 ns     15150049 ns           46 TracksPropagated=4.32579M/s
CUDA sync propagation/8        1579443 ns      1574180 ns          402 TracksPropagated=40.6561k/s
CUDA sync propagation/16       1916695 ns      1910796 ns          367 TracksPropagated=133.976k/s
CUDA sync propagation/32       2423741 ns      2416539 ns          290 TracksPropagated=423.747k/s
CUDA sync propagation/64       2790960 ns      2782183 ns          251 TracksPropagated=1.47223M/s
CUDA sync propagation/128      4954248 ns      4942169 ns          151 TracksPropagated=3.31514M/s
CUDA sync propagation/256     15537559 ns     15505154 ns           45 TracksPropagated=4.22672M/s
```

Reply:
Ok, correction, there was a bug in the PR. There does seem to be a slight difference:

Sorting on theta sector
```Running ./bin/detray_benchmark_cuda_array
Run on (48 X 1797.74 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x24)
  L1 Instruction 32 KiB (x24)
  L2 Unified 512 KiB (x24)
  L3 Unified 32768 KiB (x4)
Load Average: 1.11, 0.98, 1.35
--------------------------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------
CUDA unsync propagation/8      1810918 ns      1805887 ns          346 TracksPropagated=35.4396k/s
CUDA unsync propagation/16     2196059 ns      2190089 ns          321 TracksPropagated=116.89k/s
CUDA unsync propagation/32     2255072 ns      2248088 ns          311 TracksPropagated=455.498k/s
CUDA unsync propagation/64     2835461 ns      2826132 ns          248 TracksPropagated=1.44933M/s
CUDA unsync propagation/128    3910987 ns      3901016 ns          192 TracksPropagated=4.19993M/s
CUDA unsync propagation/256   12241997 ns     12215791 ns           58 TracksPropagated=5.36486M/s
CUDA sync propagation/8        1834123 ns      1829156 ns          353 TracksPropagated=34.9888k/s
CUDA sync propagation/16       2210750 ns      2204672 ns          320 TracksPropagated=116.117k/s
CUDA sync propagation/32       2264046 ns      2257410 ns          310 TracksPropagated=453.617k/s
CUDA sync propagation/64       2891798 ns      2883179 ns          243 TracksPropagated=1.42065M/s
CUDA sync propagation/128      3994973 ns      3984976 ns          186 TracksPropagated=4.11144M/s
CUDA sync propagation/256     12808931 ns     12781180 ns           55 TracksPropagated=5.12754M/s```


Sorting on theta directly
```Running ./bin/detray_benchmark_cuda_array
Run on (48 X 1797.6 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x24)
  L1 Instruction 32 KiB (x24)
  L2 Unified 512 KiB (x24)
  L3 Unified 32768 KiB (x4)
Load Average: 1.56, 0.98, 1.40
--------------------------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------
CUDA unsync propagation/8      1718087 ns      1713267 ns          367 TracksPropagated=37.3555k/s
CUDA unsync propagation/16     2233579 ns      2227494 ns          315 TracksPropagated=114.927k/s
CUDA unsync propagation/32     2298005 ns      2291165 ns          306 TracksPropagated=446.934k/s
CUDA unsync propagation/64     2697623 ns      2689106 ns          260 TracksPropagated=1.52318M/s
CUDA unsync propagation/128    4019020 ns      4008956 ns          187 TracksPropagated=4.08685M/s
CUDA unsync propagation/256   12336555 ns     12309757 ns           57 TracksPropagated=5.32391M/s
CUDA sync propagation/8        1717021 ns      1712189 ns          372 TracksPropagated=37.379k/s
CUDA sync propagation/16       2233491 ns      2227255 ns          317 TracksPropagated=114.94k/s
CUDA sync propagation/32       2304860 ns      2297486 ns          305 TracksPropagated=445.705k/s
CUDA sync propagation/64       2737803 ns      2728854 ns          257 TracksPropagated=1.501M/s
CUDA sync propagation/128      4113359 ns      4102710 ns          182 TracksPropagated=3.99346M/s
CUDA sync propagation/256     12920203 ns     12892396 ns           54 TracksPropagated=5.08331M/s```